lib/PuppeteerSharp.Tests/Page/EvaluateTests.cs

```diff
@@ -0,0 +1,166 @@
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                var exception = await Assert.ThrowsAsync<EvaluationFailedException>(() =>
+                {
+                    return page.EvaluateAsync<object>("() => not.existing.object.property");
+                });
+
+                Assert.NotNull(exception);
+                Assert.True(exception.Message.Contains("not is not defined"));
+            }
+        }
+
+        [Fact]
+        public async Task SouldReturnComplexObjects()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic obj = new
+                {
+                    foo = "bar!"
+                };
+                dynamic result = await page.EvaluateAsync("a => a", obj);
+                Assert.Equal("bar!", result.foo.ToString());
+            }
+        }
+
+        [Fact]
+        public async Task ShouldReturnNaN()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<dynamic>("() => NaN");
+                Assert.Equal(double.NaN, result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldReturnNegative0()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<dynamic>("() => -0");
+                Assert.Equal(-0, result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldReturnInfinity()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<dynamic>("() => Infinity");
+                Assert.Equal(double.PositiveInfinity, result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldReturnNegativeInfinty()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<dynamic>("() => -Infinity");
+                Assert.Equal(double.NegativeInfinity, result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldAcceptNullAsOneOfMultipleParameters()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<bool>("(a, b) => Object.is(a, null) && Object.is(b, 'foo')", null, "foo");
+                Assert.True(result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldProperlySerializeNullFields()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateAsync<dynamic>("() => ({a: undefined})");
+                Assert.Null(result.a);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldAcceptAString()
```

Comment:
not sure this is relevant to our project since argument is always a string

Reply:
Agreed